### PR TITLE
Added ability to dynamically start and stop timeline programmatically

### LINKED ISSUE #2008: Is there a way to dynamically turn off/on the timeline without killing application ?
Can we turn off the timeline profiler by setting env config. Currently, if I have to profile, I turn the profiler on it writes huge amount of logs and then I have to kill my application. 
Rather than killing application, I would like to dynamically turn it off, by setting a flag or config in a file. 

**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

### PR DESCRIPTION
**Pull request by tgaddair**
This PR introduces a new Python API for starting and stopping the timeline:

```
hvd.start_timeline('/path/to/timeline.json', mark_cycles=True)
...
hvd.stop_timeline()
```

The user may still choose to start the timeline via the command line and then use `hvd.stop_timeline()` to stop it programmatically after some number of steps.

Fixes #2008.

### CODE REVIEW

**Comment by Vikas-kum** [comment]
hi @tgaddair , 
Would you like to update this PR. Changes looks good and it would be really helpful to merge this in.

**Comment by Vikas-kum** [comment]
Just checking back on this. We would really appreciate if this feature can be merged. Not sure if I can provide any help but please do let me know if I can help with anything. @tgaddair

**Comment by Vikas-kum** [comment]
I think the error is happening in start_timeline method. 

```
test_torch.py::TorchTests::test_timeline_api test_torch.py::TorchTests::test_timeline_api [25a3916b2fa9:2891] *** An error occurred in MPI_Allreduce
[25a3916b2fa9:2891] *** reported by process [2819358721,0]
[25a3916b2fa9:2891] *** on communicator MPI COMMUNICATOR 3 DUP FROM 0
[25a3916b2fa9:2891] *** MPI_ERR_OTHER: known error not in list
[25a3916b2fa9:2891] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[25a3916b2fa9:2891] ***    and potentially your MPI job)
[25a3916b2fa9:2891] [0] func:/usr/local/lib/libopen-pal.so.40(opal_backtrace_buffer+0x20) [0x7fc3bd295090]
[25a3916b2fa9:2891] [1] func:/usr/local/lib/libmpi.so.40(ompi_mpi_abort+0x258) [0x7fc3bda484f8]
[25a3916b2fa9:2891] [2] func:/usr/local/lib/libmpi.so.40(ompi_mpi_errors_are_fatal_comm_handler+0xc9) [0x7fc3bda38ed9]
[25a3916b2fa9:2891] [3] func:/usr/local/lib/libmpi.so.40(ompi_errhandler_invoke+0x115) [0x7fc3bda38305]
[25a3916b2fa9:2891] [4] func:/usr/local/lib/libmpi.so.40(PMPI_Allreduce+0x2ee) [0x7fc3bda5cf5e]
[25a3916b2fa9:2891] [5] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(_ZN7horovod6common13MPIController19CrossRankBitwiseAndERSt6vectorIxSaIxEEi+0x2c) [0x7fc3b47c432c]
[25a3916b2fa9:2891] [6] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(_ZN7horovod6common16CacheCoordinator4syncESt10shared_ptrINS0_10ControllerEEb+0x1af) [0x7fc3b478c43f]
[25a3916b2fa9:2891] [7] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(_ZN7horovod6common10Controller23CoordinateCacheAndStateERNS0_16CacheCoordinatorE+0x6f) [0x7fc3b475637f]
[25a3916b2fa9:2891] [8] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(_ZN7horovod6common10Controller19ComputeResponseListERSt6atomicIbERNS0_18HorovodGlobalStateE+0x23b8) [0x7fc3b475d888]
[25a3916b2fa9:2891] [9] func:/usr/local/lib/python3.6/dist-packages/horovod/torch/mpi_lib_v2.cpython-36m-x86_64-linux-gnu.so(+0x4e1a9) [0x7fc3b47771a9]
[25a3916b2fa9:2891] [10] func:/usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so(+0x5790f) [0x7fc3c2bb390f]
[25a3916b2fa9:2891] [11] func:/lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7fc3c99036db]
[25a3916b2fa9:2891] [12] func:/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7fc3c9c3ca3f]

```

File: horovod/common/operations.cc [hunk 1]
```diff
@@ -694,6 +694,36 @@ bool horovod_is_initialized() {
...
 }
 
+bool horovod_start_timeline(const char* file_name, bool mark_cycles) {
+  if (!horovod_global.initialization_done) {
+    return false;
+  }
+
+  bool is_coordinator = horovod_global.controller->IsCoordinator();
+  if (is_coordinator) {
+    horovod_global.timeline.Initialize(file_name, horovod_global.controller->GetSize());
```
Comment (Vikas-kum) [inline horovod/common/operations.cc:704]: i guess for safety you should check if not inited then init. 
will this cause timeline to get reinited if starttimeline is called twice ?

File: horovod/common/operations.cc [hunk 2]
```diff
@@ -694,6 +694,36 @@ bool horovod_is_initialized() {
...
+  horovod_global.controller->SetTimelineEnabled(true);
+  return true;
+}
+
+bool horovod_stop_timeline() {
+  if (!horovod_global.initialization_done) {
+    return false;
+  }
+
+  horovod_global.controller->SetTimelineEnabled(false);
```
Comment (Vikas-kum) [inline horovod/common/operations.cc:717]: i guess you need to make this function(SetTimelineEnabled) thread safe. As there would be 2 different processes accessing this variable. One process trying to write the value while at the same time other process can be reading the timeline_enabled_

https://github.com/horovod/horovod/blob/e4554de96100f5a0e8686cd41cf99a6fe8a71e62/horovod/common/controller.cc#L661 and L 670 in same file

Comment (Vikas-kum) [reply horovod/common/operations.cc:717]: I also think that you should start timeline when horovod is inited with SetTimelineEnabled is false if env is not set. 

This way you would just need to protect the race for timeline_Enabled_ which can be concurrently accessed.

**Comment by tgaddair** [comment]
Hey @Vikas-kum, thanks for the review.  Yes, the reason this hasn't been merged is due to the race condition you identified.  Basically, we just need to put a big mutex around the `BackgroundThreadLoop` and only start/stop the timeline during the "sleep" between loops.

I'd like to get this merged in soon, but unfortunately I have other things that are taking priority right now.  If you're interested, please feel free to commandeer the PR.

**Comment by Vikas-kum** [comment]
> Hey @Vikas-kum, thanks for the review. Yes, the reason this hasn't been merged is due to the race condition you identified. Basically, we just need to put a big mutex around the `BackgroundThreadLoop` and only start/stop the timeline during the "sleep" between loops.
> 
> I'd like to get this merged in soon, but unfortunately I have other things that are taking priority right now. If you're interested, please feel free to commandeer the PR.

@tgaddair Thanks. I have made commit over your branch and created PR https://github.com/horovod/horovod/pull/2215 to address the issues of races.  Would you be able to please take a look at the PR.

**Pull request closed without merging**